### PR TITLE
Bugfix/course/archive/various fixes

### PR DESCRIPTION
**Pull request by ivanchimeno**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I implemented the changes with a good performance and prevented too many database calls

### Motivation and Context

The current implementation of the course archive has some issues:
- Various temporary directories weren't deleted after they were no longer needed
- An exception was thrown when a programming exercise did not have any repository URLs
- Unnecessary logging polluted the logs
- The date format used throughout the zip files included slashes which can be problematic
-  The general structure of the zip files wasn't optimal

### Description

This PR resolved the issues mentioned above.

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Generate some exercises (supported are: programming, text, file upload, and modeling)
4. Participate in these exercises with various users. The aim is to trigger some submissions for theses exercises
5. Set the course end date to the past so that it's over 
6. Navigate to the course detail screen and press the Archive Course button
7. Wait until archiving is complete and download the archive
8. The archive name should be {Course-Short-Name}-{Course-Title}-{Timestamp}.zip
9. The zipped files inside the archive should be {Course-Short-Name}-{Exercise-Title}-{Timestamp}.zip
10. Also make sure that the student submissions are included in the zip files. Programming exercises also have the template, solution, and test repositories

### Test Coverage

SubmissionExportService 91%
FileService 83%
CourseExportService 67%
ProgrammingExerciseExportService 42%
GitService 74%

### CODE REVIEW

**Review by airelawaleria — COMMENTED** [review]
Everything is archived correctly(for all students and exercises). The only thing is that the timestamp for me was one hour late. The image was taken exactly after the archiving process was finished so the one hour time difference is visible. Firstly, i though maybe  it is the issue with my browser settings, so after Chrome it tried it with microsoft edge and the timestamps didn't work at all (second screenshot - should be 12.04). Still maybe the issue with the browser, but i don't know, so let me know if that's the case :)
image
image

**Comment by ivanchimeno** [comment]
> Everything is archived correctly(for all students and exercises). The only thing is that the timestamp for me was one hour late. The image was taken exactly after the archiving process was finished so the one hour time difference is visible. Firstly, i though maybe it is the issue with my browser settings, so after Chrome it tried it with microsoft edge and the timestamps didn't work at all (second screenshot - should be 12.04). Still maybe the issue with the browser, but i don't know, so let me know if that's the case :)
> image
> image

Good catch! I accidentally added a UTC offset which is now removed

**Review by JohannesWeiss — DISMISSED** [review]
Worked as expected on ts1 (using the previous build), but with the offsets in time. This commit should fix it, hence approve!

**Review by airelawaleria — DISMISSED** [review]

**Review by sebastiantpeter — COMMENTED** [review]
Worked well on ts2. However I am not 100% sure if the archiving works, if e.g. one programming exercise and a text exercise have exactly the same title. Would this matter?

**Review by philgabr — CHANGES_REQUESTED** [review]
In general, this works fine :+1:

However, I created named 3 different exercises exactely the same (2 text ex. & 1 model ex., all named "text"), and it only created only one respective folder:
image

File: src/main/java/de/tum/in/www1/artemis/service/CourseExportService.java [hunk 1]
```diff
@@ -90,28 +92,38 @@ public void setExamService(ExamService examService) {
...
+            notifyUserAboutCourseExportState(course.getId(), CourseExportState.COMPLETED, "");
+        }
     }
 
     /**
      * Exports all exercises of the course and adds them into the directory
      * outputDir/exercises/
      *
-     * @param course       The course where the exercises are located
+     * @param course       The course where the exercises are locat
```
Comment (nicolasruscher) [inline src/main/java/de/tum/in/www1/artemis/service/CourseExportService.java:121]: I think this is a typo

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseExportService.java [hunk 2]
```diff
@@ -132,8 +130,8 @@ public Path exportProgrammingExercise(ProgrammingExercise exercise, String pathT
 
         try {
             // Zip the student and instructor repos together.
-            // The filename of the zip is: {Exercise_Short_Name}-{Exercise_Title}.zip
-            var filename = exercise.getShortName() + "-" + exercise.getTitle() + ".zip";
+            var timestamp = ZonedDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd-Hmss"));
+            var filename = exercise.getCourseViaExerciseGroupOrCourseMember().getShortName() + "-" + exercise.getTitle() + "-" + timestamp + ".zip";
```
Comment (nicolasruscher) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseExportService.java:134]: I think this causes some of the problems described with the same archive titles. Maybe adding a prefix for the different exercise types would solve the problem?

**Review by nicolasruscher — APPROVED** [review]

**Review by nicolasruscher — CHANGES_REQUESTED** [review]
See my comments above, I clicked the wrong button to approve the PR, sorry 😁

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseExportService.java (hunk 2 above)
Comment (ivanchimeno) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseExportService.java]: Wow thanks for the catch! I append the exercise id to the zip file name now in order to make each file unique

**Comment by ivanchimeno** [comment]
> In general, this works fine +1
> 
> However, I created named 3 different exercises exactely the same (2 text ex. & 1 model ex., all named "text"), and it only created only one respective folder:
> image

Thanks for the catch! I append the exercise id to the filename now in order to differentiate exercises that have the same title.

**Review by philgabr — COMMENTED** [review]
I wanted to test the new build (8) & created the resp. release. 
Somehow, it seems like Artemis cannot start up correctly again (I get an endless "Artemis is starting up..." screen, tested on ts1 & ts3)
image

**Review by philgabr — APPROVED** [review]
Works fine after merge :+1:

image

**Review by airelawaleria — APPROVED** [review]

**Review by JohannesWeiss — APPROVED** [review]

**Review by EneaGore — APPROVED** [review]

**Review by krusche — APPROVED** [review]

**Pull request merged**